src/shapes/space_filler/SpaceFiller.js

```diff
@@ -0,0 +1,175 @@
+        // will crash from recursion, so we'll set a ceiling where needed
+        if (fillers[changes.fillerType]) {
+          changes.fillerOrder = Math.min(attrs.fillerOrder || 1, fillers[changes.fillerType].maxOrder || 7)
+        }
+        return changes
+      }
+    },
+  }
+}
+
+export default class SpaceFiller extends Shape {
+  constructor() {
+    super('Space Filler')
+    this.linkText = 'Fractal charm: space filling curves'
+    this.link = 'https://www.youtube.com/watch?v=RU0wScIj36o'
+  }
+
+  getInitialState() {
+    return {
+      ...super.getInitialState(),
+      ...{
+        type: 'space_filler',
+        selectGroup: 'Erasers',
+        repeatEnabled: false,
+        canTransform: false,
+        startingSize: 1,
+        canChangeSize: false,
+        fillerOrder: 6,
+        fillerType: 'Hilbert'
+      }
+    }
+  }
+
+  // Algorithm adapted from https://editor.p5js.org/simontiger/full/2CrT12N4
+  getVertices(state) {
+    const machine = state.machine
+    const shape = state.shape
+    const order = shape.fillerOrder || 1
+
+    let sizeX, sizeY
+    if (machine.rectangular) {
+      sizeX = machine.maxX - machine.minX
+      sizeY = machine.maxY - machine.minY
+    } else {
+      sizeX = sizeY = machine.maxRadius * 2.0
+    }
+
+    let config = fillers[shape.fillerType]
+    if (shape.fillerType === 'Morton') {
+      // generate our vertices using space filling curve recursion
+      config.sizeX = sizeX
+      config.sizeY = sizeY
+      return this.spaceFillingCurve(order, config)
+    } else {
+      // generate our vertices using a set of l-system rules
+      let config = fillers[shape.fillerType]
+      config.steps = order
+
+      if (config.side === undefined) { config.side = 5 }
+      if (config.angle === undefined) { config.angle = Math.PI/2 }
+
+      let curve = lsystemPath(lsystem(config), config)
+      let scale = config.orderGrow ? order : 1
+
+      return resizeVertices(curve, sizeX*scale, sizeY*scale)
+    }
+  }
+
+  getOptions() {
+    return options
+  }
+
+  spaceFillingCurve(order, config) {
+    // create initial seed vertices that are sized to fill the entire machine area
+    let size = Math.max(config.sizeX, config.sizeY) / config.grid
+    let seedPoints = []
+    let seedPointsOrdered = []
+
+    for (let j = 0; j < config.grid; j++) {
+      for (let i = 0; i < config.grid; i++) {
+        seedPoints.push(new Victor(i*size + size/2, j*size + size/2))
+      }
+    }
+
+    for (let i=0; i<config.seed.length; i++) {
+      seedPointsOrdered.push(seedPoints[config.seed[i]])
+    }
+
+    config.seedPointsOrdered = seedPointsOrdered
+    config.size = size
+
+    // recursively generate our curve
+    let curve = this.fillSpace(order, config)
+    return curve.map(vertex => vertex.add({x: -size*config.grid/2, y: -size*config.grid/2}))
+  }
+
+  // Algorithm adapted from https://editor.p5js.org/simontiger/full/2CrT12N4
```

Comment:
Is there any way to contact this simontiger and make sure they are ok with us publishing it with an MIT license? I doubt they will care, but I like to give them the opportunity to say yes.

Reply:
> How hard would it be to make it always start or end in a corner/perimeter or center? Do any of these end in the center?
I think they all start/end on the perimeter.

Reply:
We only use his code for the Morton curve. I'll either reach out, find another option, or remove it.